### PR TITLE
D44105_D44104_D44103 RQM BugFix related to dynamic filed mapping

### PR DESCRIPTION
**Pull request by prashelke**
Thank you for your contribution.
Before submitting this PR, please make sure:

- [x] PR description and commit message should describe the changes done in this PR
- [x] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [x] Latest Code from master or specific release branch is merged to your branch
- [x] No unwanted\commented\junk code is included
- [ ] No new warning upon build solution
- [ ] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [x] Sanity Tests are successfully executed for New and Existing Functionality
- [x] Verify that changes are compatible with all relevant browsers and platforms.
- [x] After creating pull request there should not be any conflicts
- [x] Resolve all Codacy comments
- [x] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Update the Help Library document to match any feature changes

## Summary by CodeRabbit

- **New Features**
	- Enhanced export functionality to ensure mandatory fields are validated before exporting to RQM.
	- Improved retrieval and handling of custom attributes and external fields during import and export processes.
	- Introduced a method to retrieve external item fields based on ALM configuration.

- **Bug Fixes**
	- Enhanced error handling and logging for mandatory fields in the export process.

- **Refactor**
	- Simplified code structure by encapsulating logic for retrieving external fields into dedicated methods, improving maintainability.

### CODE REVIEW

File: Ginger/Ginger/ALM/Repository/RQMRepository.cs [hunk 1]
```diff
@@ -324,6 +291,46 @@ public override bool ExportBusinessFlowToALM(BusinessFlow businessFlow, bool per
...
+                    originalExternalFields.Add(fieldToAdd);
+                }
+            }
+            else
+            {
+                originalExternalFields = WorkSpace.Instance.Solution.ExternalItemsFields;
+            }
+
+            return originalExternalFields;
+        }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/ALM/Repository/RQMRepository.cs]: Done

File: Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs [hunk 2]
```diff
@@ -1429,17 +1430,20 @@
...
                             CustomAttributeName = CustomAttributeListing.GetElementsByTagName("ns4:title").Item(0).InnerText;
                             CustomAttributeItemType = CustomAttributeListing.GetElementsByTagName("ns2:scope").Item(0).InnerText;
+                            CustomAttributefieldType = CustomAttributeListing.GetElementsByTagName("ns2:type").Item(0).InnerText;
                             CustomAttributeMandatory = "false";
 
                             itemfield.ItemType = CustomAttributeItemType;
                             itemfield.ID = CustomAttributeID;
                             itemfield.TypeIdentifier = typeIdentifier;
                             itemfield.Name = CustomAttributeName;
+                            itemfield.Type = CustomAttributefieldType;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs:1449]: No need

File: Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs [hunk 3]
```diff
@@ -1248,17 +1248,21 @@
                 //step 2. Get all the custom attribute link as per the response
                 //setp 3. Get all Custom Attribute details by API 
                 //step 4. Get All custom Attribute Possible values
-
-                GetCustomAttributes(bw, fields, rqmSserverUrl, loginData, ref baseUri_, ref selfLink_, ref maxPageNumber_);
+                ObservableList<ExternalItemFieldBase> Customfields = new ObservableList<ExternalItemFieldBase>();
+                Customfields = GetCustomAttributes(bw, rqmSserverUrl, loginData, ref baseUri_, ref selfLink_, ref maxPageNumber_);
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs]: Done

File: Ginger/GingerCoreNET/ALMLib/RQM/ExportToRQM.cs [hunk 4]
```diff
@@ -683,6 +692,45 @@ private string CreateExecutionRecord(string bfExportedID, ActivitiesGroup activG
...
+                        ItemType = externalItemField.ItemType,
+                        Guid = externalItemField.Guid,
+                        IsCustomField = externalItemField.IsCustomField,
+                        SelectedValue = existingField != null && !string.IsNullOrEmpty(existingField.SelectedValue)
+                            ? existingField.SelectedValue
+                            : externalItemField.SelectedValue
+                    };
+
+                    originalExternalFields.Add(fieldToAdd);
+                }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ALMLib/RQM/ExportToRQM.cs]: updated

**Review by Maheshkale447 — APPROVED** [review]

**Pull request merged**